Panic in etcd gateway start

### What happened?

A panic happens when running `etcd gateway start --discovery-srv someplace.com --listen-addr=127.0.5.6:23790 --trusted-ca-file somepath/ca.pem` in v3.5.2 where no panic happens when running it in v3.4.18.

### What did you expect to happen?

No panic, establish connection without error or fail with error message.

### How can we reproduce it (as minimally and precisely as possible)?

The above command on Windows would do the trick.
I can give you the CA file and the DNS name privately. Just tell me where to send the information.

### Anything else we need to know?

* When specifying `--insecure-discovery` the panic does not occur and things operate normally.
* The problem does not occur in v3.4.18
* The cluster etcd version is v3.4.18
* I tried setting `GODEBUG=x509sha1=1,x509ignoreCN=0,http2debug=2` and combinations thereof - no change and not further diagnostic output. Some of those options prob. even aren't valid anymore.

### Etcd version (please run commands below)

<details>

```console
$ etcd --version
etcd Version: 3.5.2
Git SHA: 99018a77b
Go Version: go1.16.3
Go OS/Arch: windows/amd64

$ etcdctl version
etcdctl version: 3.5.2
API version: 3.5
```

</details>


### Etcd configuration (command line flags or environment variables)

<details>

gateway start --discovery-srv someplace.com --listen-addr=127.0.5.6:23790 --trusted-ca-file somepath/ca.pem

</details>


### Etcd debug information (please run commands blow, feel free to obfuscate the IP address or FQDN in the output)

<details>

```console
$ etcdctl member list -w table
+------------------+---------+-------+----------------------------------+----------------------------------+------------+
|        ID        | STATUS  | NAME  |          PEER ADDRS              |         CLIENT ADDRS             | IS LEARNER |
+------------------+---------+-------+----------------------------------+----------------------------------+------------+
| 237a24d9080da0db | started | etcd4 | https://etcd4.someplace.com:2380 | https://etcd4.someplace.com:2379 |      false |
| 42b03093b37f16a9 | started | etcd5 | https://etcd5.someplace.com:2380 | https://etcd5.someplace.com:2379 |      false |
| 6aea52aba38d03f1 | started | etcd3 | https://etcd3.someplace.com:2380 | https://etcd3.someplace.com:2379 |      false |
| a8cf2d8a635f6a96 | started | etcd2 | https://etcd2.someplace.com:2380 | https://etcd2.someplace.com:2379 |      false |
| f5f365ee667c3149 | started | etcd1 | https://etcd1.someplace.com:2380 | https://etcd1.someplace.com:2379 |      false |
+------------------+---------+-------+----------------------------------+----------------------------------+------------+
$ etcdctl --endpoints=<member list> endpoint status -w table
+----------------------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+--------+
|           ENDPOINT               |        ID        | VERSION | DB SIZE | IS LEADER | IS LEARNER | RAFT TERM | RAFT INDEX | RAFT APPLIED INDEX | ERRORS |
+----------------------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+--------+
| https://etcd1.someplace.com:2379 | f5f365ee667c3149 |  3.4.18 |  9.5 MB |      true |      false |     11648 |   82909472 |           82909472 |        |
| https://etcd2.someplace.com:2379 | a8cf2d8a635f6a96 |  3.4.18 |  9.5 MB |     false |      false |     11648 |   82909472 |           82909472 |        |
| https://etcd3.someplace.com:2379 | 6aea52aba38d03f1 |  3.4.18 |  9.5 MB |     false |      false |     11648 |   82909472 |           82909472 |        |
| https://etcd4.someplace.com:2379 | 237a24d9080da0db |  3.4.18 |  9.5 MB |     false |      false |     11648 |   82909472 |           82909472 |        |
| https://etcd5.someplace.com:2379 | 42b03093b37f16a9 |  3.4.18 |  9.5 MB |     false |      false |     11648 |   82909472 |           82909472 |        |
+----------------------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+--------+
```

</details>


### Relevant log output

```shell
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x0 pc=0x5920d9]

goroutine 1 [running]:
go.etcd.io/etcd/client/pkg/v3/transport.ValidateSecureEndpoints(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc000038180, 0x17, ...)
        /tmp/etcd-release-3.5.2/etcd/release/etcd/client/pkg/transport/tls.go:37 +0x1d9
go.etcd.io/etcd/server/v3/etcdmain.discoverEndpoints(0xc00045a280, 0xc00003c100, 0xa, 0xc000038180, 0x17, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /tmp/etcd-release-3.5.2/etcd/release/etcd/server/etcdmain/util.go:63 +0x250
go.etcd.io/etcd/server/v3/etcdmain.startGateway(0xc000455180, 0xc00045a230, 0x0, 0x5)
        /tmp/etcd-release-3.5.2/etcd/release/etcd/server/etcdmain/gateway.go:105 +0x2a5
github.com/spf13/cobra.(*Command).execute(0xc000455180, 0xc00045a1e0, 0x5, 0x5, 0xc000455180, 0xc00045a1e0)
        /usr/local/google/home/siarkowicz/.gvm/pkgsets/go1.16.3/global/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:856 +0x2c2
github.com/spf13/cobra.(*Command).ExecuteC(0x187ac60, 0x60, 0xc0000a8000, 0xc00008ff48)
        /usr/local/google/home/siarkowicz/.gvm/pkgsets/go1.16.3/global/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:960 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
        /usr/local/google/home/siarkowicz/.gvm/pkgsets/go1.16.3/global/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:897
go.etcd.io/etcd/server/v3/etcdmain.Main(0xc0000dc000, 0x8, 0x8)
        /tmp/etcd-release-3.5.2/etcd/release/etcd/server/etcdmain/main.go:32 +0x85
main.main()
        /tmp/etcd-release-3.5.2/etcd/release/etcd/server/main.go:32 +0x4c
```

--- a/client/pkg/transport/tls.go
+++ b/client/pkg/transport/tls.go
@@ -28,6 +28,8 @@ func ValidateSecureEndpoints(tlsInfo TLSInfo, eps []string) ([]string, error) {
 	if err != nil {
 		return nil, err
 	}
+	defer t.CloseIdleConnections()
+
 	var errs []string
 	var endpoints []string
 	for _, ep := range eps {

--- a/client/pkg/transport/tls.go
+++ b/client/pkg/transport/tls.go
@@ -15,6 +15,7 @@
 package transport
 
 import (
+	"context"
 	"fmt"
 	"strings"
 	"time"
@@ -34,7 +35,7 @@ func ValidateSecureEndpoints(tlsInfo TLSInfo, eps []string) ([]string, error) {
 			errs = append(errs, fmt.Sprintf("%q is insecure", ep))
 			continue
 		}
-		conn, cerr := t.Dial("tcp", ep[len("https://"):])
+		conn, cerr := t.DialContext(context.Background(), "tcp", ep[len("https://"):])
 		if cerr != nil {
 			errs = append(errs, fmt.Sprintf("%q failed to dial (%v)", ep, cerr))
 			continue

--- a/client/pkg/transport/tls_test.go
+++ b/client/pkg/transport/tls_test.go
@@ -78,16 +78,11 @@ func TestValidateSecureEndpoints(t *testing.T) {
 		t.Run(name, func(t *testing.T) {
 			secureEps, err := ValidateSecureEndpoints(*tlsInfo, test.endPoints)
 			if test.expectedErr != (err != nil) {
-				t.Errorf("Unexpected error, got: %v, want: %v", err, test.expectedError)
+				t.Errorf("Unexpected error, got: %v, want: %v", err, test.expectedErr)
 			}
 
-			if err == nil && !test.expectedErr {
-				if len(secureEps) != len(test.expectedEndpoints) {
-					t.Errorf("expected %v endpoints, got %v", len(test.expectedEndpoints), len(secureEps))
-				}
-				if !reflect.DeepEqual(test.expectedEndpoints, secureEps) {
-					t.Errorf("expected endpoints %v, got %v", test.expectedEndpoints, secureEps)
-				}
+			if !reflect.DeepEqual(test.expectedEndpoints, secureEps) {
+				t.Errorf("expected endpoints %v, got %v", test.expectedEndpoints, secureEps)
 			}
 		})
 	}

--- a/client/pkg/transport/tls_test.go
+++ b/client/pkg/transport/tls_test.go
@@ -17,7 +17,7 @@ package transport
 import (
 	"net/http"
 	"net/http/httptest"
-	"strings"
+	"reflect"
 	"testing"
 )
 
@@ -33,18 +33,64 @@ func TestValidateSecureEndpoints(t *testing.T) {
 	srv := httptest.NewServer(http.HandlerFunc(remoteAddr))
 	defer srv.Close()
 
-	insecureEps := []string{
-		"http://" + srv.Listener.Addr().String(),
-		"invalid remote address",
+	tests := map[string]struct {
+		endPoints         []string
+		expectedEndpoints []string
+		expectedErr       bool
+	}{
+		"invalidEndPoints": {
+			endPoints: []string{
+				"invalid endpoint",
+			},
+			expectedEndpoints: nil,
+			expectedErr:       true,
+		},
+		"insecureEndpoints": {
+			endPoints: []string{
+				"http://127.0.0.1:8000",
+				"http://" + srv.Listener.Addr().String(),
+			},
+			expectedEndpoints: nil,
+			expectedErr:       true,
+		},
+		"secureEndPoints": {
+			endPoints: []string{
+				"https://" + srv.Listener.Addr().String(),
+			},
+			expectedEndpoints: []string{
+				"https://" + srv.Listener.Addr().String(),
+			},
+			expectedErr: false,
+		},
+		"mixEndPoints": {
+			endPoints: []string{
+				"https://" + srv.Listener.Addr().String(),
+				"http://" + srv.Listener.Addr().String(),
+				"invalid end points",
+			},
+			expectedEndpoints: []string{
+				"https://" + srv.Listener.Addr().String(),
+			},
+			expectedErr: true,
+		},
 	}
-	if _, err := ValidateSecureEndpoints(*tlsInfo, insecureEps); err == nil || !strings.Contains(err.Error(), "is insecure") {
-		t.Error("validate secure endpoints should fail")
-	}
-
-	secureEps := []string{
-		"https://" + srv.Listener.Addr().String(),
-	}
-	if _, err := ValidateSecureEndpoints(*tlsInfo, secureEps); err != nil {
-		t.Error("validate secure endpoints should succeed")
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			secureEps, err := ValidateSecureEndpoints(*tlsInfo, test.endPoints)
+			if test.expectedErr && err == nil {
+				t.Errorf("expected error")
+			}
+			if !test.expectedErr && err != nil {
+				t.Errorf("unexpected error: %v", err)
+			}
+			if err == nil && !test.expectedErr {
+				if len(secureEps) != len(test.expectedEndpoints) {
+					t.Errorf("expected %v endpoints, got %v", len(test.expectedEndpoints), len(secureEps))
+				}
+				if !reflect.DeepEqual(test.expectedEndpoints, secureEps) {
+					t.Errorf("expected endpoints %v, got %v", test.expectedEndpoints, secureEps)
+				}
+			}
+		})
 	}
 }

--- a/client/pkg/transport/tls_test.go
+++ b/client/pkg/transport/tls_test.go
@@ -1,3 +1,17 @@
+// Copyright 2022 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
 package transport
 
 import (

--- a/client/pkg/transport/tls_test.go
+++ b/client/pkg/transport/tls_test.go
@@ -77,12 +77,10 @@ func TestValidateSecureEndpoints(t *testing.T) {
 	for name, test := range tests {
 		t.Run(name, func(t *testing.T) {
 			secureEps, err := ValidateSecureEndpoints(*tlsInfo, test.endPoints)
-			if test.expectedErr && err == nil {
-				t.Errorf("expected error")
+			if test.expectedErr != (err != nil) {
+				t.Errorf("Unexpected error, got: %v, want: %v", err, test.expectedError)
 			}
-			if !test.expectedErr && err != nil {
-				t.Errorf("unexpected error: %v", err)
-			}
+
 			if err == nil && !test.expectedErr {
 				if len(secureEps) != len(test.expectedEndpoints) {
 					t.Errorf("expected %v endpoints, got %v", len(test.expectedEndpoints), len(secureEps))

--- a/client/pkg/transport/tls_test.go
+++ b/client/pkg/transport/tls_test.go
@@ -1,0 +1,36 @@
+package transport
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+)
+
+func TestValidateSecureEndpoints(t *testing.T) {
+	tlsInfo, err := createSelfCert(t)
+	if err != nil {
+		t.Fatalf("unable to create cert: %v", err)
+	}
+
+	remoteAddr := func(w http.ResponseWriter, r *http.Request) {
+		w.Write([]byte(r.RemoteAddr))
+	}
+	srv := httptest.NewServer(http.HandlerFunc(remoteAddr))
+	defer srv.Close()
+
+	insecureEps := []string{
+		"http://" + srv.Listener.Addr().String(),
+		"invalid remote address",
+	}
+	if _, err := ValidateSecureEndpoints(*tlsInfo, insecureEps); err == nil || !strings.Contains(err.Error(), "is insecure") {
+		t.Error("validate secure endpoints should fail")
+	}
+
+	secureEps := []string{
+		"https://" + srv.Listener.Addr().String(),
+	}
+	if _, err := ValidateSecureEndpoints(*tlsInfo, secureEps); err != nil {
+		t.Error("validate secure endpoints should succeed")
+	}
+}

